### PR TITLE
Autogenstudio Updates [CSV support, Workflow Export, Skill Editing, Windows Testing ]

### LINKED ISSUE #1442: [Issue]: Editing a Skill
### Describe the issue

In Studio, to change a skill, I need to delete and add it again with the new code.
Is that by design or did I mess up the installation?
If it is the current state of things, are there plans to add the capability to change the code of a skill through the studio interface?
Is there a work around?

### Steps to reproduce

_No response_

### Screenshots and logs

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by victordibia**
## Why are these changes needed?

This PR adds improvements to AutoGen Studio 

- Fix a bug with workflow specification #1428 . Thanks @junkei-okinawa 
- Add Support for viewing CSV files in AutoGen Studio #1355. Thanks @pretherford 
- Fix Install Guide for Windows Users #1470. Thanks @kripper 
- Other Features
   - Add Models as high level object with support for testing models in UI. #1404
   - Support for downloading workflows (sanitized to remove api keys)  #1287
   - Add Support for editing skills with a full fledged code editor  . Fixes #1442 
   - Add speaker selection to Group Chat UI #1373 
   - Add temperate slider for models and default_auto_reply in UI #1521 

## Related issue number 

#1373  #1442 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by gagb** [comment]
I get the following error when I try to build this PR with the studio dev container.
Complains about a product env being missing.

[creation.log](https://github.com/microsoft/autogen/files/14105591/creation.log)

**Comment by kripper** [comment]
> Complains about a product env being missing.

It's asking to create a `.env.production` file:
https://github.com/microsoft/autogen/tree/autogenstudio/samples/apps/autogen-studio/frontend#setting-env-variables-for-the-ui

Should we use the default values for production?
To solve, we can a) add this file to the repo or b) use the default values if the file is missing.

Please also see: https://github.com/microsoft/autogen/issues/1465#issuecomment-1917968452

**Review by junkei-okinawa — APPROVED** [review]
I have verified that the edits are reflected and the functionality works.
- [x] Fix a bug with workflow specification Fixed a bug in Workflow Specification Modal state management #1428 . 
- [x] Add Support for viewing CSV files in AutoGen Studio Autogenstudio Add CSV file support #1355. 
- [x] Fix Install Guide for Windows Users Fix build issues #1470. 
* Other Features
- [x] Add Models as high level object with support for testing models in UI. [Feature Request]: "Test" button for Models #1404
- [x] Support for downloading workflows (sanitized to remove api keys) [Autogen Studio Feature Request]: The ability to export agents and workflows as python code. #1287
-> @victordibia  ~~To copy a workflow, click on the "copy" icon, then click on "Cancel" in the "Workflow Specification" modal that appears. in the "Workflow Specification" modal that appears after clicking the "copy" icon. This will be the next issue to be corrected.~~
-> My apologies. It seems I made an operational error. I withdraw my point as it does not reproduce.
- [x] Add Support for editing skills with a full fledged code editor . Fixes [Issue]: Editing a Skill #1442
- [x] Add speaker selection to Group Chat UI [Feature Request]: Ability to Predefine Speaker Flow in Studio #1373

**Comment by cheng-tan** [comment]
small issue: In csv report, if select showing 10 items or 20 items per page, it still shows 50 items per page

File: samples/apps/autogen-studio/README.md [hunk 1]
```diff
@@ -137,6 +144,13 @@ A: Yes, you can view the generated messages in the debug console of the web UI,
 Yes. AutoGen standardizes on the openai model api format, and you can use any api server that offers an openai compliant endpoint. In the AutoGen Studio UI, each agent has an `llm_config` field where you can input your model endpoint details including `model name`, `api key`, `base url`, `model type` and `api version`. For Azure OpenAI models, you can find these details in the Azure portal. Note that for Azure OpenAI, the `model name` is the deployment id or engine, and the `model type` is "azure".
 For other OSS models, we recommend using a server such as vllm to instantiate an openai compliant endpoint.
 
+**Q: The Server Starts But I Can't Access the UI**
```
Comment (cheng-tan) [inline samples/apps/autogen-studio/README.md:147]: nit: the other Questions are all in lower case

File: samples/apps/autogen-studio/autogenstudio/utils/dbdefaults.json [hunk 2]
```diff
@@ -202,7 +184,7 @@
                 },
                 "human_input_mode": "NEVER",
                 "max_consecutive_auto_reply": 8,
-                "system_message": "You are a helpful assistant that can suggest a travel itinerary for a user. You are the primary coordinator who will receive suggestions or advice from other agents (local_assistant, language_assistant). You must ensure that the finally plan integrates the suggestions from other agents or team members. YOUR FINAL RESPONSE MUST BE THE COMPLETE PLAN that ends with the word TERMINATE. "
+                "system_message": "You are a helpful assistant that can suggest a travel itinerary for a user. You are the primary cordinator who will receive suggestions or advice from other agents (local_assistant, language_assistant). You must ensure that the finally plan integrates the suggestions from other agents or team members. YOUR FINAL RESPONSE MUST BE THE COMPLETE PLAN that ends with the word TERMINATE. "
```
Comment (cheng-tan) [inline samples/apps/autogen-studio/autogenstudio/utils/dbdefaults.json:187]: coordinator

File: samples/apps/autogen-studio/autogenstudio/workflowmanager.py [hunk 3]
```diff
@@ -113,24 +113,42 @@ def sanitize_agent_spec(self, agent_spec: AgentFlowSpec) -> AgentFlowSpec:
...
-            agent_spec.config.system_message = (
-                autogen.AssistantAgent.DEFAULT_SYSTEM_MESSAGE
-                + "\n\n"
-                + agent_spec.config.system_message
-                + "\n\n"
-                + skills_prompt
-            )
+        if agent_spec.skills:
+            # get skill prompt, also write skills to a file named skills.py
+            skills_prompt = ""
```
Comment (cheng-tan) [inline samples/apps/autogen-studio/autogenstudio/workflowmanager.py:145]: nit: remove this line

File: samples/apps/autogen-studio/frontend/src/components/atoms.tsx [hunk 4]
```diff
@@ -627,14 +631,14 @@ export const ModelSelector = ({
...
 
-  const sanitizeModelConfig = (config: IModelConfig) => {
-    const sanitizedConfig: IModelConfig = { model: config.model };
-    if (config.api_key) sanitizedConfig.api_key = config.api_key;
-    if (config.base_url) sanitizedConfig.base_url = config.base_url;
-    if (config.api_type) sanitizedConfig.api_type = config.api_type;
-    if (config.api_version) sanitizedConfig.api_version = config.api_version;
-    return sanitizedConfig;
-  };
+  // const sanitizeModelConfig = (config: IModelConfig) => {
```
Comment (cheng-tan) [inline samples/apps/autogen-studio/frontend/src/components/atoms.tsx:634]: nit: remove this

File: samples/apps/autogen-studio/frontend/src/components/atoms.tsx [hunk 5]
```diff
@@ -684,31 +688,71 @@ export const ModelSelector = ({
...
+            label: <>No models found</>,
             value: 0,
           },
         ];
 
   const modelOnClick: MenuProps["onClick"] = ({ key }) => {
     const selectedIndex = parseInt(key.toString());
     let selectedModel = models[selectedIndex];
-    selectedModel = sanitizeModelConfig(selectedModel);
+    selectedModel = selectedModel;
```
Comment (cheng-tan) [inline samples/apps/autogen-studio/frontend/src/components/atoms.tsx:712]: nit: remove

File: samples/apps/autogen-studio/frontend/src/components/atoms.tsx [hunk 6]
```diff
@@ -722,7 +766,7 @@ export const ModelSelector = ({
 
   const handleOk = () => {
     if (newModelConfig?.model.trim()) {
-      const sanitizedConfig = sanitizeModelConfig(newModelConfig);
+      const sanitizedConfig = newModelConfig;
```
Comment (cheng-tan) [inline samples/apps/autogen-studio/frontend/src/components/atoms.tsx:769]: just want to confirm `newModelConfig` has been sanitized?

File: samples/apps/autogen-studio/frontend/src/components/utils.ts [hunk 7]
```diff
@@ -278,7 +298,7 @@ export const sampleWorkflowConfig = (type = "twoagents") => {
   const groupChatAssistantConfig = Object.assign({}, assistantConfig);
   groupChatAssistantConfig.name = "groupchat_assistant";
   groupChatAssistantConfig.system_message =
-    "You are a helpful assistant skilled at coordinating a group of other assistants to solve a task. ";
+    "You are a helpful assistant skilled at cordinating a group of other assistants to solve a task. ";
```
Comment (cheng-tan) [inline samples/apps/autogen-studio/frontend/src/components/utils.ts:301]: coordinating

File: samples/apps/autogen-studio/notebooks/groupchat_spec.json [hunk 8]
```diff
@@ -56,7 +56,7 @@
             },
             "human_input_mode": "NEVER",
             "max_consecutive_auto_reply": 8,
-            "system_message": "You are a helpful assistant that can suggest a travel itinerary for a user. You are the primary coordinator who will receive suggestions or advice from other agents (local_assistant, language_assistant). You must ensure that the finally plan integrates the suggestions from other agents or team members. YOUR FINAL RESPONSE MUST BE THE COMPLETE PLAN that ends with the word TERMINATE. "
+            "system_message": "You are a helpful assistant that can suggest a travel itinerary for a user. You are the primary cordinator who will receive suggestions or advice from other agents (local_assistant, language_assistant). You must ensure that the finally plan integrates the suggestions from other agents or team members. YOUR FINAL RESPONSE MUST BE THE COMPLETE PLAN that ends with the word TERMINATE. "
```
Comment (cheng-tan) [inline samples/apps/autogen-studio/notebooks/groupchat_spec.json:59]: coordinator

**Review by cheng-tan — APPROVED** [review]

File: samples/apps/autogen-studio/frontend/src/components/atoms.tsx (hunk 6 above)
Comment (victordibia) [reply samples/apps/autogen-studio/frontend/src/components/atoms.tsx:769]: Oh yes .. model sanitization logic has been moved to the backend, in workflowmanager.
Here, sanitization just means we remove any extra fields from the model object (e.g. things like id, timestamp, description etc).
This enforces sanitizated behavior at the backend. 
That is why sanitization is not needed on the frontend anymore.

**Comment by gagb** [comment]
> I get the following error when I try to build this PR with the studio dev container. Complains about a product env being missing. lmk when this is fixed and I can re-review.
> 
> [creation.log](https://github.com/microsoft/autogen/files/14105591/creation.log)

Seems like the build issues were addressed in commits after last review. Approving.

**Review by gagb — APPROVED** [review]

**Pull request merged**